src/reason_parser.mly

```diff
@@ -2491,6 +2508,121 @@ labeled_simple_pattern:
       { ("", None, $1) }
 ;
 
+jsx_arguments:
+    /* empty */ { [] }
+    | LIDENT EQUAL simple_expr jsx_arguments {
+        (* a=b *)
+        [($1, $3)] @ $4
+     }
+    | LIDENT jsx_arguments {
+        (* a (punning) *)
+        let loc_lident = mklocation $startpos($1) $endpos($1) in
+        [($1, mkexp (Pexp_ident {txt = Lident $1; loc = loc_lident}))] @ $2
+     }
+;
+
+jsx_start_tag_body:
+  | mod_longident jsx_arguments {
+     (jsx_component $1 $2, $1)
+  }
+;
+
+jsx_separate_open_close:
+  | jsx_start_tag_body GREATERLESSSLASH mod_longident {
+    (* Foo></Foo *)
+      let (component, start) = $1 in
+      let loc = mklocation $symbolstartpos $endpos in
+      let _ = ensureTagsAreEqual start $3 in
```

Comment:
Btw: When we are in "recovering parser mode", we recover from parsing failures, by replacing whole nodes with extension point nodes (which have a single string expression describing the error). In that mode, it's good because it doesn't block parsing the rest of the file, and then merlin can work its type checking recovery magic.